### PR TITLE
feat: Resend Email Integration

### LINKED ISSUE #576: [DF-387] Implement Resend support
Implement support for [https://resend.com/](https://resend.com/) as an email service provider.

Add a new `PersistedEmailProvider` type:

[https://github.com/dittofeed/dittofeed/blob/01a6db73e237897b7e961c07969ea5d60a8bcbd0/packages/isomorphic-lib/src/types.ts#L991-L994](https://github.com/dittofeed/dittofeed/blob/01a6db73e237897b7e961c07969ea5d60a8bcbd0/packages/isomorphic-lib/src/types.ts#L991-L994)

Add new `EmailServiceProviderSuccess` type:

[https://github.com/dittofeed/dittofeed/blob/01a6db73e237897b7e961c07969ea5d60a8bcbd0/packages/isomorphic-lib/src/types.ts#L1946-L1950C4](https://github.com/dittofeed/dittofeed/blob/01a6db73e237897b7e961c07969ea5d60a8bcbd0/packages/isomorphic-lib/src/types.ts#L1946-L1950C4)

Add new `EmailServiceProviderFailure` type:

[https://github.com/dittofeed/dittofeed/blob/01a6db73e237897b7e961c07969ea5d60a8bcbd0/packages/isomorphic-lib/src/types.ts#L2080-L2083](https://github.com/dittofeed/dittofeed/blob/01a6db73e237897b7e961c07969ea5d60a8bcbd0/packages/isomorphic-lib/src/types.ts#L2080-L2083)

Add new `EmailProviderSecret` type:

[https://github.com/dittofeed/dittofeed/blob/01a6db73e237897b7e961c07969ea5d60a8bcbd0/packages/isomorphic-lib/src/types.ts#L2080-L2083](https://github.com/dittofeed/dittofeed/blob/01a6db73e237897b7e961c07969ea5d60a8bcbd0/packages/isomorphic-lib/src/types.ts#L2080-L2083)A webhook for

Error handling for the `/templates/test` api endpoint:

[https://github.com/dittofeed/dittofeed/blob/01a6db73e237897b7e961c07969ea5d60a8bcbd0/packages/api/src/controllers/contentController.ts#L174-L178](https://github.com/dittofeed/dittofeed/blob/01a6db73e237897b7e961c07969ea5d60a8bcbd0/packages/api/src/controllers/contentController.ts#L174-L178)

A webhook for updating the status of sent emails:

[https://github.com/dittofeed/dittofeed/blob/01a6db73e237897b7e961c07969ea5d60a8bcbd0/packages/api/src/controllers/webhooksController.ts#L23-L24](https://github.com/dittofeed/dittofeed/blob/01a6db73e237897b7e961c07969ea5d60a8bcbd0/packages/api/src/controllers/webhooksController.ts#L23-L24)

Handle the email provider type when sending messages:

[https://github.com/dittofeed/dittofeed/blob/01a6db73e237897b7e961c07969ea5d60a8bcbd0/packages/backend-lib/src/messageTemplates.ts#L479-L481](https://github.com/dittofeed/dittofeed/blob/01a6db73e237897b7e961c07969ea5d60a8bcbd0/packages/backend-lib/src/messageTemplates.ts#L479-L481)

Create the Resend email provider:

[https://github.com/dittofeed/dittofeed/blob/01a6db73e237897b7e961c07969ea5d60a8bcbd0/packages/dashboard/src/lib/email.ts#L86-L100](https://github.com/dittofeed/dittofeed/blob/01a6db73e237897b7e961c07969ea5d60a8bcbd0/packages/dashboard/src/lib/email.ts#L86-L100)

Add Place in Settings to configure email provider:

[https://github.com/dittofeed/dittofeed/blob/01a6db73e237897b7e961c07969ea5d60a8bcbd0/packages/dashboard/src/pages/settings.page.tsx#L847-L854](https://github.com/dittofeed/dittofeed/blob/01a6db73e237897b7e961c07969ea5d60a8bcbd0/packages/dashboard/src/pages/settings.page.tsx#L847-L854)

<sub>From [SyncLinear.com](https://synclinear.com) | [DF-387](https://linear.app/dittofeed/issue/DF-387/implement-resend-support)</sub>

### PR DESCRIPTION
**Pull request by neo773**
This PR adds Resend Email Integration

/closes #576
/claim #576

### CODE REVIEW

**Comment by maxgurewitz** [comment]
Wow, so fast @neo773 ! I should be able to review this by EOD tomorrow.

File: packages/api/package.json [hunk 1]
```diff
@@ -43,6 +43,8 @@
     "neverthrow": "^6.0.0",
     "openapi-types": "^12.1.0",
     "remeda": "^1.12.0",
+    "resend": "^2.1.0",
+    "svix": "^1.15.0",
```
Comment (maxgurewitz) [inline packages/api/package.json:47]: We need to implement this without introducing a svix dependency.

File: packages/backend-lib/src/destinations/resend.ts [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+      "workspaceId",
+      "journeyId",
+      "runId",
+      "messageId",
+      "userId",
+      "templateId"
+    ])
+  );
+
+  switch (event) {
```
Comment (maxgurewitz) [inline packages/backend-lib/src/destinations/resend.ts:92]: Great!

File: packages/backend-lib/src/journeys/userWorkflow/activities.ts [hunk 3]
```diff
@@ -677,6 +678,54 @@ async function sendEmailWithPayload(
             replyTo,
           });
         }
+        case EmailProviderType.Resend: {
```
Comment (maxgurewitz) [inline packages/backend-lib/src/journeys/userWorkflow/activities.ts:681]: This is actually deprecated code. Feel free to replace this with a simple `throw` statement here. It'll be deprecated soon.

**Review by maxgurewitz — CHANGES_REQUESTED** [review]

**Comment by maxgurewitz** [comment]
Just tested the code. Seems to be working! 🥳 Fix the webhook dependency and we should be good :)

File: packages/api/package.json (hunk 1 above)
Comment (neo773) [reply packages/api/package.json:47]: image

SVIX (Webhooks as a Service) recommends using their library to verify the payloads and also I spoke to Resend's team they said the same thing

Let me know if you still want to proceed with verifying it manually.

**Comment by maxgurewitz** [comment]
@neo773 you're so close! Is there anyway we could push this through before the end of the weekend? If not I may have to take over.

**Comment by neo773** [comment]
@maxgurewitz 
Hi,
I left a reply to your comment on the webhook code, seems like you may have missed it in your inbox https://github.com/dittofeed/dittofeed/pull/584#discussion_r1457338199

**Comment by maxgurewitz** [comment]
@neo773 responded! Also feel free to join our [discord](https://discord.gg/HajPkCG4Mm) if you'd like to chat with lower friction!

File: packages/api/src/controllers/webhooksController.ts [hunk 4]
```diff
@@ -108,6 +111,107 @@ export default async function webhookController(fastify: FastifyInstance) {
...
+          error: "Missing secret.",
+        });
+      }
+
+      if (!request.rawBody || typeof request.rawBody !== "string") {
+        logger().error({ workspaceId }, "Missing rawBody on resend webhook.");
+        return reply.status(500).send();
+      }
+
+      const svixId = request.headers["svix-id"];
```
Comment (maxgurewitz) [inline packages/api/src/controllers/webhooksController.ts:177]: You've still got some svix references in your code!

**Review by maxgurewitz — CHANGES_REQUESTED** [review]
Can you remove the remaining svix code, and test the webhook with ngrok or a similar tool to make sure it's working? Thank you!

**Review by maxgurewitz — APPROVED** [review]

**Pull request merged**